### PR TITLE
[perceval] Add Askbot backend

### PR DESCRIPTION
**Pull request by albertinisg**
This is the first version of the Askbot backend. Still some improvements are needed:

- Fix the comments retrieval using the AJAX call
- Add cache

### CODE REVIEW

File: perceval/backends/askbot.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+        return html_question_items
+
+
+class AskbotClient:
+    """Askbot client.
+
+    This class implements a simple client to retrieve distinct
+    kind of data from an Askbot site.
+
+    :param base_url: URL of the Askbot site
```
Comment (sduenas) [inline perceval/backends/askbot.py]: The list of parameters does not match with the signature of `__init__`.

File: perceval/backends/askbot.py [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+
+        self.url = self.parsed_args.url
+        self.from_date = str_to_datetime(self.parsed_args.from_date)
+        self.tag = self.parsed_args.tag
+        self.outfile = self.parsed_args.outfile
+
+        self.backend = Askbot(self.url, tag=self.tag)
+
+    def run(self):
+        """Fetch and print the bugs.
```
Comment (sduenas) [inline perceval/backends/askbot.py]: Check the documentation strings. There are wrong references to bugs.

File: perceval/backends/askbot.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+                    'page': page,
+                    'sort': self.ORDER_API
+                 }
+        response = self.__call(path, params)
+        return response
+
+    def get_html_question(self, question_id, page=1):
+        """Retrieve a raw HTML question and all it's information.
+
+        :param question_id: question identifier
```
Comment (sduenas) [inline perceval/backends/askbot.py]: I miss here the description of the `page` parameter.

File: perceval/backends/askbot.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+    """Askbot backend.
+
+    This class retrieves the questions posted in an Askbot site.
+    To initialize this class the URL must be provided. The `url`
+    will be set as the origin of the data.
+
+    :param url: Askbot site URL
+    :param origin: identifier of the repository; when `None` or an
+        empty string are given, it will be set to `url` value
+    """
```
Comment (sduenas) [inline perceval/backends/askbot.py]: I miss `tag` description here.

Comment (sduenas) [inline perceval/backends/askbot.py]: `origin` is not given as parameter.

File: perceval/backends/askbot.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+
+        :param from_date: obtain questions/answers updated since this date
+
+        :returns: a generator of items
+        """
+
+        if not from_date:
+            from_date = DEFAULT_DATETIME
+
+        from_date = from_date.timestamp()
```
Comment (sduenas) [inline perceval/backends/askbot.py]: You have to call first to `datetime_to_utc(from_date)` to convert the date to UTC.

File: perceval/backends/askbot.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+
+        :returns: this backend supports items resuming
+        """
+        return True
+
+    @classmethod
+    def has_caching(cls):
+        """Returns whether it supports caching items on the fetch process.
+
+        :returns: this backend supports items cache
```
Comment (sduenas) [inline perceval/backends/askbot.py]: This backend doesn't supports items cache

File: perceval/backends/askbot.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+        """
+        return False
+
+    def fetch_html_question(self, question):
```
Comment (sduenas) [inline perceval/backends/askbot.py]: This should be a private method and I think it should be renamed to `__fetch_question`.

File: perceval/backends/askbot.py [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+import logging
+
+from ..backend import Backend, metadata, BackendCommand
+from ..utils import urljoin, DEFAULT_DATETIME, str_to_datetime
+
+logger = logging.getLogger(__name__)
+
+
+class AskbotParser:
+
```
Comment (sduenas) [inline perceval/backends/askbot.py]: I miss some documentation here.

File: perceval/backends/askbot.py [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+    def parse_question(self, question, html_question):
```
Comment (sduenas) [inline perceval/backends/askbot.py]: In my opinion, this method should be changed to just parse a HTML question page. Then, in the backend build the item with all the parts (answers, comments, etc). I think the same behaviour must be followed by all the parsing methods.

**Comment by albertinisg** [comment]
First comments for the first review addressed. Going for the second round (I still cannot see them)

File: perceval/backends/askbot.py [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+
+        for page in range(2, len(html_question)+1):
+            position = page - 1
+            try:
+                html_question[position]
+            except IndexError:
+                continue
+            else:
+                bs_question = bs4.BeautifulSoup(html_question[position], "html.parser")
+                answers.extend(AskbotParser.parse_answers(bs_question))
```
Comment (sduenas) [inline perceval/backends/askbot.py]: This call to BeautifulSoup should be part of the parser, not the backend. The backend doesn't need to know about how the parsing is done.

File: perceval/backends/askbot.py [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+
+        html_question_items = []
+
+        npages = 1
+        next_request = True
+
+        while next_request:
+            html_question = self.client.get_html_question(question['id'], npages)
+            html_question_items.append(html_question)
+            bs_question = bs4.BeautifulSoup(html_question, "html.parser")
```
Comment (sduenas) [inline perceval/backends/askbot.py]: BeautifulSoup call should be part of the parser.

File: perceval/_version.py [hunk 12]
```diff
@@ -1,2 +1,2 @@
 # Versions compliant with PEP 440 https://www.python.org/dev/peps/pep-0440
-__version__ = "0.4.2.dev0"
+__version__ = "0.4.3.dev0"
```
Comment (sduenas) [inline perceval/_version.py]: To avoid problems with version, would you mind to revert this change?

**Comment by albertinisg** [comment]
@sduenas changes requested addressed!
